### PR TITLE
[SPARK-42290][SQL] Fix the OOM error can't be reported when AQE on

### PR DESCRIPTION
**Pull request by Hisoka-X**
### What changes were proposed in this pull request?
When we use spark shell to submit job like this:
```scala
$ spark-shell --conf spark.driver.memory=1g

val df = spark.range(5000000).withColumn("str", lit("abcdabcdabcdabcdabasgasdfsadfasdfasdfasfasfsadfasdfsadfasdf"))
val df2 = spark.range(10).join(broadcast(df), Seq("id"), "left_outer")

df2.collect
```
This will cause the driver to hang indefinitely.
When we disable AQE, the `java.lang.OutOfMemoryError` will be throws.

After I check the code, the reason are wrong way to use `Throwable::initCause`. It happened when OOM be throw on https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala#L184 . Then https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala#L2401 will be executed. 
It use `new SparkException(..., case=oe).initCause(oe.getCause)`. 
The doc in `Throwable::initCause` say
```
This method can be called at most once. It is generally called from within the constructor,
or immediately after creating the throwable. If this throwable was created with Throwable(Throwable)
or Throwable(String, Throwable), this method cannot be called even once.
```
So when we call it, the `IllegalStateException` will be throw. Finally, the `promise.tryFailure(ex)` never be called. The driver will be blocked.

### Why are the changes needed?
Fix the OOM never be reported bug

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add new test

### CODE REVIEW

**Comment by Hisoka-X** [comment]
cc @cloud-fan @itholic

File: sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala [hunk 1]
```diff
@@ -308,6 +308,10 @@ class QueryExecutionErrorsSuite
     }
   }
 
+  test("SPARK-42290: NotEnoughMemory error can't be create") {
```
Comment (LuciferYang) [inline sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala:311]: Yeah,  `OutOfMemoryError` has no with `cause` constructor.

**Review by LuciferYang — APPROVED** [review]
+1, LGTM

**Comment by LuciferYang** [comment]
cc @MaxGekk FYI

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for reporting and fix, @Hisoka-X .

According to the JIRA, v3.4.0 is this only one affected?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 2]
```diff
@@ -2398,7 +2398,7 @@ private[sql] object QueryExecutionErrors extends QueryErrorsBase {
         "autoBroadcastjoinThreshold" -> SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key,
         "driverMemory" -> SparkLauncher.DRIVER_MEMORY,
         "analyzeTblMsg" -> analyzeTblMsg),
-      cause = oe).initCause(oe.getCause)
+      cause = oe.getCause)
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:2401]: Got it. SPARK-40663 is the root cause.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

I also verified that this doesn't exist in Apache Spark 3.3.2.
```
scala> val df = spark.range(5000000).withColumn("str", lit("abcdabcdabcdabcdabasgasdfsadfasdfasdfasfasfsadfasdfsadfasdf"))
df: org.apache.spark.sql.DataFrame = [id: bigint, str: string]

scala> val df2 = spark.range(10).join(broadcast(df), Seq("id"), "left_outer")
df2: org.apache.spark.sql.DataFrame = [id: bigint, str: string]

scala> df2.collect
java.lang.OutOfMemoryError: Not enough memory to build and broadcast the table to all worker nodes. As a workaround, you can either disable broadcast by setting spark.sql.autoBroadcastJoinThreshold to -1 or increase the spark driver memory by setting spark.driver.memory to a higher value.
  at org.apache.spark.sql.errors.QueryExecutionErrors$.notEnoughMemoryToBuildAndBroadcastTableError(QueryExecutionErrors.scala:1838)
  at org.apache.spark.sql.execution.exchange.BroadcastExchangeExec.$anonfun$relationFuture$1(BroadcastExchangeExec.scala:183)
  at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withThreadLocalCaptured$1(SQLExecution.scala:191)
  at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
  at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
  at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
  at java.base/java.lang.Thread.run(Thread.java:829)

scala> sc.version
res1: String = 3.3.2
```

**Review by kazuyukitanimura — APPROVED** [review]
LGTM (non-binding)

**Comment by dongjoon-hyun** [comment]
I verified manually. Merged to master/3.4.
```
$ build/sbt "sql/testOnly *.QueryExecutionErrorsSuite -- -z SPARK-42290"
[info] QueryExecutionErrorsSuite:
13:10:15.573 WARN org.apache.hadoop.util.NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
[info] - SPARK-42290: NotEnoughMemory error can't be create (211 milliseconds)
13:10:16.741 WARN org.apache.spark.sql.errors.QueryExecutionErrorsSuite:

===== POSSIBLE THREAD LEAK IN SUITE o.a.s.sql.errors.QueryExecutionErrorsSuite, threads: rpc-boss-3-1 (daemon=true), shuffle-boss-6-1 (daemon=true) =====
[info] Run completed in 2 seconds, 415 milliseconds.
[info] Total number of tests run: 1
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 1, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[success] Total time: 22 s, completed Jun 8, 2023, 1:10:16 PM
```

**Comment by dongjoon-hyun** [comment]
Thank you, @Hisoka-X , @LuciferYang , @kazuyukitanimura .

**Comment by cloud-fan** [comment]
late LGTM

**Comment by itholic** [comment]
last LGTM. Thanks for the fix

**Pull request closed without merging**